source/Dgraph/Client/DgraphClient.cs

```diff
@@ -14,119 +14,111 @@
-            return next;
+            return DgraphExecute(
+                async (dg) =>
+                {
+                    await dg.LoginAsync(new LoginRequest
+                    {
+                        Userid = user,
+                        Password = password,
+                        Namespace = ns
+                    }, options ?? new CallOptions());
+                    return Result.Ok();
+                },
+                (rpcEx) => Result.Fail(new ExceptionalError(rpcEx))
+            );
         }
 
-        public async Task<FluentResults.Result> Alter(Api.Operation op, CallOptions? options = null)
+        Task<Result> IDgraphClient.Alter(Api.Operation op, CallOptions? options)
         {
-            return await DgraphExecute(
+            return DgraphExecute(
                 async (dg) =>
                 {
                     await dg.AlterAsync(op, options ?? new CallOptions());
-                    return Results.Ok();
+                    return Result.Ok();
                 },
-                (rpcEx) => Results.Fail(new FluentResults.ExceptionalError(rpcEx))
+                (rpcEx) => Result.Fail(new ExceptionalError(rpcEx))
             );
         }
 
-        public async Task<FluentResults.Result<string>> CheckVersion(CallOptions? options = null)
+        Task<Result<string>> IDgraphClient.CheckVersion(CallOptions? options)
         {
-            return await DgraphExecute(
+            return DgraphExecute(
                 async (dg) =>
                 {
                     var versionResult = await dg.CheckVersionAsync(new Check(), options ?? new CallOptions());
-                    return Results.Ok<string>(versionResult.Tag); ;
+                    return Result.Ok<string>(versionResult.Tag); ;
                 },
-                (rpcEx) => Results.Fail<string>(new FluentResults.ExceptionalError(rpcEx))
+                (rpcEx) => Result.Fail<string>(new ExceptionalError(rpcEx))
             );
         }
 
-        public async Task<FluentResults.Result> Login(Api.LoginRequest lr, CallOptions? options = null)
+        ITransaction IDgraphClient.NewTransaction()
         {
-            return await DgraphExecute(
-                async (dg) =>
-                {
-                    await dg.LoginAsync(lr, options ?? new CallOptions());
-                    return Results.Ok();
-                },
-                (rpcEx) => Results.Fail(new FluentResults.ExceptionalError(rpcEx))
-            );
+            AssertNotDisposed();
+            return new Transaction(client: this);
+        }
+
+        IQuery IDgraphClient.NewReadOnlyTransaction(bool bestEffort)
+        {
+            AssertNotDisposed();
+            return new Transaction(client: this, readOnly: true, bestEffort: bestEffort);
+        }
+
+        #endregion
+
+        #region execution
+
+        private int NextConnection = 0;
+        private int GetNextConnection()
+        {
+            var next = NextConnection;
```

Comment:
do we need this assignment?

Reply:
And will this be thread safe?

Reply:
There is a small chance that the counter will skip an increment occasionally if this function is called from multiple threads. I think thread safety would only be a concern if the generated protobuf client `Api.Dgraph.DgraphClient` is not thread safe, in which case we'd have to introduce some kind of locking on it, but I don't see this happening in dgo.

For the counter, I could replace the increment of `NextConnection` with a random selection, eg:

```c#
private Api.Dgraph.DgraphClient GetNextClient()
{
    // Randomly pick the next client to use.
    return dgraphs[Random.Shared.Next(dgraphs.Count)];
}
```

`Random.Shared` is thread safe. 

@mangalaman93 any thoughts on this?

Reply:
If you want to keep the more basic increment logic, there are ways to atomically increment an integer as well:

```c#
private uint NextClientIndex = 0;
private Api.Dgraph.DgraphClient GetNextClient()
{
    // Randomly pick the next client to use.
    return dgraphs[(int)(Interlocked.Increment(ref NextClientIndex) % dgraphs.Count)];
}
```

Reply:
atomic increment or random works. We use random in dgo, but either looks equally good to me. You can use what is more lightweight in C#.

Reply:
dgo client is thread safe internally. For example, it is possible that client.Login is called in parallel and trying to login to two different namespace in parallel. Such calls will be executed in sequence through a lock. Check that something like this won't do bad things in C#.

Reply:
Ok, I used the random selection logic. I will have to look at the generated code from the protobufs; there are no locking mechanisms in the wrapping code here.

Reply:
Ok no, the client.Login here is not thread safe. Fixing Login & JWTs is out of scope for Phase 1.

Reply:
Just file a bug for now on GitHub. Thanks

Reply:
Seems like the Issues tab is disabled in this repo for me.

Reply:
The issue tab is disabled for everyone... need some admin to reactivate it.

Reply:
working on it.